daemon/libnetwork/ipams/defaultipam/address_space_test.go

```diff
@@ -498,3 +498,119 @@ func TestPoolAllocateAndRelease(t *testing.T) {
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 2, 2) },
+				func(tcs testClosures) { tcs.alloc(netip.MustParseAddr("2001:db8::10"), 3, 3) },
+				func(tcs testClosures) { tcs.release(netip.MustParseAddr("2001:db8::10"), 2, 2) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 3, 3) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 4, 4) },
+			},
+		},
+		{
+			name: "check allocated IP counters with IP Range for IPv6",
+			predefined: []*ipamutils.NetworkToSplit{
+				{Base: netip.MustParsePrefix("2001:db8::/56"), Size: 64},
+			},
+			subnet:  netip.MustParsePrefix("2001:db8::/120"),
+			ipRange: netip.MustParsePrefix("2001:db8::80/125"),
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 2, 1) },
+				func(tcs testClosures) { tcs.alloc(netip.MustParseAddr("2001:db8::10"), 3, 1) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 4, 2) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 5, 3) },
+				func(tcs testClosures) { tcs.release(netip.MustParseAddr("2001:db8::10"), 4, 3) },
+				func(tcs testClosures) { tcs.alloc(netip.MustParseAddr("2001:db8::85"), 5, 4) },
+				func(tcs testClosures) { tcs.release(netip.MustParseAddr("2001:db8::85"), 4, 3) },
+			},
+		},
+		{
+			name: "check allocated IP counters without IP Range",
+			predefined: []*ipamutils.NetworkToSplit{
+				{Base: netip.MustParsePrefix("10.0.0.0/16"), Size: 24},
+			},
+			subnet: netip.MustParsePrefix("10.0.0.0/24"),
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 3, 3) },
+				func(tcs testClosures) { tcs.alloc(netip.MustParseAddr("10.0.0.10"), 4, 4) },
+				func(tcs testClosures) { tcs.release(netip.MustParseAddr("10.0.0.10"), 3, 3) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 4, 4) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 5, 5) },
+			},
+		},
+		{
+			name: "check allocated IP counters with IP Range",
+			predefined: []*ipamutils.NetworkToSplit{
+				{Base: netip.MustParsePrefix("10.0.0.0/16"), Size: 24},
+			},
+			subnet:  netip.MustParsePrefix("10.0.0.0/24"),
+			ipRange: netip.MustParsePrefix("10.0.0.128/29"),
+			calls: []func(tcs testClosures){
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 3, 1) },
+				func(tcs testClosures) { tcs.alloc(netip.MustParseAddr("10.0.0.10"), 4, 1) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 5, 2) },
+				func(tcs testClosures) { tcs.alloc(netip.Addr{}, 6, 3) },
+				func(tcs testClosures) { tcs.release(netip.MustParseAddr("10.0.0.10"), 5, 3) },
+				func(tcs testClosures) { tcs.alloc(netip.MustParseAddr("10.0.0.131"), 6, 4) },
+				func(tcs testClosures) { tcs.release(netip.MustParseAddr("10.0.0.131"), 5, 3) },
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			as, err := newAddrSpace(tc.predefined)
+			assert.NilError(t, err)
+
+			err = as.allocateSubnet(tc.subnet, tc.ipRange)
```

Comment:
Where are you testing the cases with multiple ip-range allocations under a single subnet? Be sure to cover both the disjoint ranges (e.g. `--subnet 192.168.0.0/24 --ip-range 192.168.0.0/25`, `--subnet 192.168.0.0/24 --ip-range 192.168.0.128/25`) [and overlapping ranges](https://github.com/moby/moby/issues/46756) (e.g. `--subnet 192.168.0.0/24 --ip-range 192.168.0.0/25`, `--subnet 192.168.0.0/24 --ip-range 192.168.0.64/26`) cases